sw/airborne/firmwares/rotorcraft/oneloop/oneloop_andi.c

```diff
@@ -0,0 +1,1524 @@
+    andi_du[i] = (float)(andi_du_n[i] * ratio_u_un[i]);
+  }
+
+  if (volando) {
+    // Add the increments to the actuators
+    float_vect_sum(andi_u, actuator_state_1l, andi_du, ANDI_NUM_ACT);
+    andi_u[ANDI_NUM_ACT]   = andi_du[ANDI_NUM_ACT]   + att_1l[0];
+    andi_u[ANDI_NUM_ACT+1] = andi_du[ANDI_NUM_ACT+1] + att_1l[1];
+  } else {
+    // Not in flight, so don't increment
+    float_vect_copy(andi_u, andi_du, ANDI_NUM_ACT);
+    andi_u[ANDI_NUM_ACT] = andi_du[ANDI_NUM_ACT];
+    andi_u[ANDI_NUM_ACT+1] = andi_du[ANDI_NUM_ACT+1];
+  }
+
+  // TODO : USE THE PROVIDED MAX AND MIN and change limits for phi and theta
+  // Bound the inputs to the actuators
+  for (i = 0; i < ANDI_NUM_ACT_TOT; i++) {
+    Bound(andi_u[i], act_min[i], act_max[i]);
+  }
+
+  /*Commit the actuator command*/
+  stabilization_cmd[COMMAND_THRUST] = 0;
+  for (i = 0; i < ANDI_NUM_ACT; i++) {
+    actuators_pprz[i] = (int16_t) andi_u[i];
+    stabilization_cmd[COMMAND_THRUST] += actuator_state_1l[i];
+  }
+  stabilization_cmd[COMMAND_THRUST] = stabilization_cmd[COMMAND_THRUST]/4.0;
+  if(autopilot.mode==AP_MODE_ATTITUDE_DIRECT){
+    eulers_zxy_des.phi   =  andi_u[ANDI_NUM_ACT];
+    eulers_zxy_des.theta =  andi_u[ANDI_NUM_ACT+1];
+  } else {
+    eulers_zxy_des.phi   =  andi_u[ANDI_NUM_ACT];
+    eulers_zxy_des.theta =  andi_u[ANDI_NUM_ACT+1];
+  }
+  psi_des_deg = psi_des_rad * 180.0 / M_PI;  
+  printf("###################################################\n");
+}
+
+/** @brief  Function to reconstruct actuator state using first order dynamics */
+void get_act_state_oneloop(void)
+{
+  int8_t i;
+  float prev_actuator_state_1l;
+  for (i = 0; i < ANDI_NUM_ACT; i++) {
+    prev_actuator_state_1l = actuator_state_1l[i];
+    actuator_state_1l[i] = prev_actuator_state_1l + act_dynamics_d[i] * (andi_u[i] - prev_actuator_state_1l);
+    if(!autopilot_get_motors_on()){
+      actuator_state_1l[i] = 0.0;
+    }
+    Bound(actuator_state_1l[i],0,MAX_PPRZ);
+  }
+}
+
+/**
+ * @brief Function that sums g1 and g2 to obtain the g1_g2 matrix. It also undoes the scaling that was done to make the values readable
+ * FIXME: make this function into a for loop to make it more adaptable to different configurations
+ */
+void sum_g1g2_1l(void) {
+  //struct FloatEulers *euler = stateGetNedToBodyEulers_f();
+  float sphi   = sinf(eulers_zxy.phi);
+  float cphi   = cosf(eulers_zxy.phi);
+  float stheta = sinf(eulers_zxy.theta);
+  float ctheta = cosf(eulers_zxy.theta);
+  float spsi   = sinf(eulers_zxy.psi);
+  float cpsi   = cosf(eulers_zxy.psi);
+  float T      = -9.81; //minus gravity is a guesstimate of the thrust force, thrust measurement would be better
+  float P      = 0.0;
+  if (ONELOOP_ANDI_AC_HAS_PUSHER){
+    P    = actuator_state_1l[ONELOOP_ANDI_PUSHER_IDX] * g1_1l[2][ONELOOP_ANDI_PUSHER_IDX] / ANDI_G_SCALING;
+  } 
+  float scaler;
+  int i = 0;
+  for (i = 0; i < ANDI_NUM_ACT_TOT; i++) {
+    // Effectiveness matrix for real actuators
```

Comment:
matrix -> vector

The meaning of 'real' actuators is unclear

Reply:
Added description. Specified that theta and phi are virtual actuators.